File: test-junit5/src/test/java/io/micronaut/test/junit5/MathServiceTestSimilarNameTest.java

package io.micronaut.test.junit5;

import io.micronaut.context.annotation.Requires;
import io.micronaut.core.util.StringUtils;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.annotation.MockBean;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import jakarta.inject.Inject;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@MicronautTest
@Requires(property = "mockito.test.enabled", defaultValue = StringUtils.FALSE, value = StringUtils.TRUE)
public class MathServiceTestSimilarNameTest {

    @Inject
    MathService mathService;


    @Test
    void testThatSimilarlyNamedTestSuitesDontLeakMocks() {
        int num = 10;
        when(mathService.compute(num))
                .then(invocation -> num * 2); // non mock impl is * 4

        final int result = mathService.compute(num);

        Assertions.assertEquals(
                20,
                result
        );
    }

    @MockBean(MathServiceImpl.class)
    MathService mathService() {
        return mock(MathService.class);
    }
}

File: test-kotlintest/src/test/kotlin/io/micronaut/test/kotlintest/MathServiceTestSimilarNameTest.kt

package io.micronaut.test.kotlintest

import io.kotlintest.shouldBe
import io.kotlintest.specs.BehaviorSpec
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import io.micronaut.test.annotation.MockBean
import io.micronaut.test.extensions.kotlintest.MicronautKotlinTestExtension.getMock
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify

@MicronautTest
class MathServiceTestSimilarNameTest(private val mathService: MathService): BehaviorSpec({

    given("test similarly named test suites dont leak mocks") {

        `when`("the mock is provided") {
            val mock = getMock(mathService)
            every { mock.compute(10) } returns 20

            then("the mock is used") {
                mock.compute(10) shouldBe 20
                verify { mock.compute(10) }
            }
        }
    }

}) {

    @MockBean(MathServiceImpl::class)
    fun mathService(): MathService {
        return mockk()
    }
}
